### PR TITLE
feat: rewrite the scanner in C

### PR DESCRIPTION
**Pull request by amaanq**

### CODE REVIEW

**Comment by vkleen** [comment]
Do you have a particular rationale for rewriting the scanner in C instead of C++?

**Comment by amaanq** [comment]
Portability and ease of compilation - it better supports projects using a pure C toolchain and the logic is simple enough to not really need the C++ containers

**Review by ErinvanderVeen — APPROVED** [review]

**Comment by amaanq** [comment]
Sure, those are valid concerns - but I've been rewriting many C++ scanners across various tree-sitter repos - for the most part, it doesn't solve a particular bug and it really is just about making compilation easier and improving portability. Though in some cases, there were bugs with leaks or buffer overflows in the original scanner that I had fixed anyways (not this one)

**Comment by amaanq** [comment]
I'll also tack on that building this for wasm is much easier than with C++ due to the need to manually specify specific std functions to export (in tree-sitter core there is a file that manually picks common functions to export)

Anyways I updated it to add static to any helper functions

**Comment by clason** [comment]
I'll also add that upstream has officially deprecated C++ scanners, and such parsers will be removed in nvim-treesitter 1.0.

**Comment by yannham** [comment]
Thanks for the information, @clason. I think this is a good argument in favor of moving the scanner to C, despite the loss of some nice abstractions.

**Comment by clason** [comment]
Upstream is adding some nice abstractions to the C headers, though, so it might not be too bad.

**Review by yannham — APPROVED** [review]
Thanks again for undertaking that. We'll probably need to rebase and regenerate the parser, as there's a conflict currently.

File: src/scanner.c [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+#include <wctype.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define VEC_RESIZE(vec, _cap)                                                  \
+    void *tmp = realloc((vec).data, (_cap) * sizeof((vec).data[0]));           \
+    assert(tmp != NULL);                                                       \
+    (vec).data = tmp;                                                          \
+    (vec).cap = (_cap);
+
```
[the remark this answers is not in the document]
Comment (amaanq) [reply src/scanner.c:13]: sorta the convention atm across grammars that leverage this mechanism, and it'll always be inlined wherever used no matter the compiler/opt level specified.

We will be providing very convenient array macros in the near feature upstream to replace this, to dedup the code and also simplify it, https://github.com/tree-sitter/tree-sitter/pull/3063 and for a demo https://github.com/tree-sitter/tree-sitter-bash/pull/255

**Comment by amaanq** [comment]
> Thanks again for undertaking that. We'll probably need to rebase and regenerate the parser, as there's a conflict currently.

C++ is a pain to get working cross-platform and in wasm especially, and there's really no need for most of its abstractions. The only used abstractions across grammars with a scanner is vectors/strings, which the near-future inclusion of an array header will provide a decent replacement for.

**Pull request merged**